system/lib/libc/dynlink.c

```diff
@@ -0,0 +1,185 @@
+  error("Invalid library handle %p", (void *)h);
+  return 1;
+}
+
+static void load_library_done(struct dso *p) {
+  // insert into linked list
+  p->prev = tail;
+  if (tail) {
+    tail->next = p;
+  }
+  tail = p;
+
+  if (!head) {
+    head = p;
+  }
+}
+
+static struct dso *load_library_start(const char *name, int flags) {
+  if (!(flags & (RTLD_LAZY | RTLD_NOW))) {
+    error("invalid mode for dlopen(): Either RTLD_LAZY or RTLD_NOW is required");
+    return NULL;
+  }
+
+  struct dso *p;
+  size_t alloc_size = sizeof *p + strlen(name) + 1;
+  p = calloc(1, alloc_size);
+  strcpy(p->name, name);
+
+  return p;
+}
+
+static void dlopen_js_onsuccess(void* handle) {
+  struct dso* p = (struct dso*)handle;
+#ifdef DYLINK_DEBUG
+  printf("dlopen_js_onsuccess: dso=%p\n", p);
+#endif
+  load_library_done(p);
+  p->onsuccess(p->user_data, p);
+}
+
+static void dlopen_js_onerror(void* handle) {
+  struct dso* p = (struct dso*)handle;
+#ifdef DYLINK_DEBUG
+  printf("dlopen_js_onsuccess: dso=%p\n", p);
+#endif
+  p->onerror(p->user_data);
+  free(p);
+}
+
+void *dlopen(const char *file, int flags) {
+  if (!file) return head;
+#ifdef DYLINK_DEBUG
+  printf("dlopen: %s [%d]\n", file, flags);
+#endif
+
+  struct dso *p;
+  int cs;
+  pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, &cs);
+  pthread_rwlock_wrlock(&lock);
+
+  /* Search for the name to see if it's already loaded */
+  for (p = head; p; p = p->next) {
+    if (!strcmp(p->name, file)) {
+#ifdef DYLINK_DEBUG
+      printf("dlopen: already opened: %p\n", p);
+#endif
+      goto end;
+    }
+  }
+
+  p = load_library_start(file, flags);
+  if (!p) goto end;
+  void* success = _dlopen_js(file, flags, p);
+  if (!success) {
+#ifdef DYLINK_DEBUG
+    printf("dlopen_js: failed\n", p);
+#endif
+    free(p);
+    p = NULL;
+    goto end;
+  }
+  load_library_done(p);
+end:
+  pthread_rwlock_unlock(&lock);
+  pthread_setcancelstate(cs, 0);
+#ifdef DYLINK_DEBUG
+    printf("dlopen_js: success: %p\n", p);
+#endif
+  return p;
+}
+
+void emscripten_dlopen(const char *filename, int flags, void* user_data, em_dlopen_callback onsuccess, em_arg_callback_func onerror) {
+  struct dso *p = load_library_start(filename, flags);
+  if (!p) {
+    onerror(user_data);
+    return;
+  }
+  p->user_data = user_data;
+  p->onsuccess = onsuccess;
+  p->onerror = onerror;
+#ifdef DYLINK_DEBUG
+  printf("calling emscripten_dlopen_js %p\n", p);
+#endif
+  _emscripten_dlopen_js(filename, flags, p, dlopen_js_onsuccess, dlopen_js_onerror);
+}
+
+__attribute__((__visibility__("hidden")))
+void *__dlsym(void *restrict p, const char *restrict s, void *restrict ra) {
+#ifdef DYLINK_DEBUG
+  printf("__dlsym dso:%p sym:%s\n", p, s);
+#endif
+  if (p != RTLD_DEFAULT && p != RTLD_NEXT && __dl_invalid_handle(p)) {
+    return 0;
+  }
+  void *res;
+  pthread_rwlock_rdlock(&lock);
+  res = _dlsym_js(p, s);
+  pthread_rwlock_unlock(&lock);
+  return res;
+}
+
+int dladdr(const void *addr, Dl_info *info) {
```

Comment:
this is one of those cases where our address space mismatches (object file/linker symbols vs fn pointers vs data address spaces) gets pretty ugly. Should we think of `addr` as just being a function pointer?

Reply:
No I think addr can be any address.. but honestly I don't think its very important for this to work.  At least this PR doesn't attempt to make it work.

Reply:
Right, I guess I meant that in a unified address space, it could be either a data or function symbol. We could probably make it work for one of those but not both (with the same function).
I agree we don't need to make it work now though. It looks like maybe we could make it return 0 (indicating failure) instead of 1?

Reply:
I guess it depends how likely its is that static data address space and function address space overlap.  I guess we can't rule it out since we want both of them to be low numbers where possible.